packages/frontend/src/hooks/useTransaction.ts

```diff
@@ -0,0 +1,238 @@
+  L1Transfer,
+  getLastLog,
+  formatLogArgs,
+  networkIdToSlug,
+} from 'src/utils'
+import { TransactionResponse, TransactionReceipt } from '@ethersproject/providers'
+import { formatEther } from '@ethersproject/units'
+import { getProviderByNetworkName } from 'src/utils/getProvider'
+import { Interface, Result } from '@ethersproject/abi'
+import { getExplorerTxUrl } from 'src/utils/getExplorerUrl'
+import { hopBridgeTokenAbi, l1BridgeAbi, l2AmmWrapperAbi } from '@hop-protocol/core/abi'
+
+export const methodToSigHashes = {
+  // L1_Bridge
+  sendToL2: '0xdeace8f5', // L1 -> L2
+  // L2_Bridge
+  swapAndSend: '0xeea0d7b2', // L2 -> L2 / L2 -> L1
+  // L1_Bridge / L2_Bridge
+  withdraw: '0x0f7aadb7', // L2 -> L2 / L2 -> L1 (bonder offline)
+}
+
+export const sigHashes = {
+  '0x095ea7b3': 'approve',
+  // L1_Bridge
+  '0xdeace8f5': 'sendToL2', // L1 -> L2
+  // L2_Bridge
+  '0xeea0d7b2': 'swapAndSend', // L2 -> L2 / L2 -> L1
+  // L1_Bridge / L2_Bridge
+  '0x0f7aadb7': 'withdraw', // L2 -> L2 / L2 -> L1 (bonder offline)
+}
+
+enum TxType {
+  approve = 'approve',
+  sendL1ToL2 = 'L1 to L2',
+  sendL2ToL2 = 'L2 to L2',
+  sendL2ToL1 = 'L2 to L1',
+  convert = 'convert',
+  wrap = 'wrap',
+  unwrap = 'unwrap',
+  addLiquidity = 'addLiquidity',
+  removeLiquidity = 'removeLiquidity',
+  stake = 'stake',
+  unstake = 'unstake',
+}
+
+interface TxObj {
+  type?: TxType
+  response: TransactionResponse
+  receipt: TransactionReceipt
+  gasCost: string
+  datetime?: string
+  methodName?: string
+  tokenSymbol?: string
+  params?: any[]
+  eventValues?: any
+  destNetworkName?: string
+  destTxHash?: string
+  destExplorerLink?: string
+}
+
+const useTransaction = (txHash?: string, slug?: string) => {
+  const [tx, setTx] = useState<any>()
+  const [completed, setCompleted] = useState<boolean>(tx?.pending === false)
+  const [networkConfirmations, setNetworkConfirmations] = useState<number>()
+  const [confirmations, setConfirmations] = useState<number>()
+  const [destCompleted, setDestCompleted] = useState<boolean>(
+    tx?.pendingDestinationConfirmation === false
+  )
+  const [loading, setLoading] = useState(false)
+  const [error, setError] = useState<any>()
+  const [txObj, setTxObj] = useState<TxObj>()
+
+  useEffect(() => {
+    async function initState() {
+      if (txHash && slug) {
+        setError(false)
+        setLoading(true)
+        try {
+          const provider = getProviderByNetworkName(slug)
+          const response = await provider.getTransaction(txHash)
+          const receipt = await provider.getTransactionReceipt(txHash)
+
+          if (!response) {
+            throw new Error(`Could not get tx on network: ${slug}`)
+          }
+
+          const { data, gasPrice } = response
+          const { gasUsed } = receipt
+          const gasCost = formatEther(gasUsed.mul(gasPrice!))
+
+          const funcSig = data.slice(0, 10)
+          const methodName = sigHashes[funcSig]
+
+          setLoading(false)
```

Comment:
would it be better to set `setLoading(false)` at the _end_ of the function once all async calls have finished?

Reply:
i experienced some delays in especially the later async calls, so i'm setting loading to false early on. additionally, if a tx was recent, or is pending being bonded, or just hasn't been mined yet, we still want to see the data that *has* been loaded so far so we can get an idea of what the current status is.